src/librustc_mir/build/expr.rs

```diff
@@ -0,0 +1,704 @@
+                    upvars.into_iter()
+                          .map(|upvar| unpack!(block = this.as_lvalue(block, upvar)))
+                          .collect();
+                block.and(Rvalue::Aggregate(AggregateKind::Closure(closure_id, substs), upvars))
+            }
+            ExprKind::Adt { adt_def, variant_index, substs, fields, base } => { // see (*) above
+                // first process the set of fields
+                let fields_map: FnvHashMap<Field<H>, Lvalue<H>> =
+                    fields.into_iter()
+                          .map(|f| (f.name, unpack!(block = this.as_lvalue(block, f.expr))))
+                          .collect();
+
+                let field_names =
+                    this.hir.fields(adt_def, variant_index);
+
+                let base =
+                    base.map(|base| unpack!(block = this.as_lvalue(block, base)));
+
+                // for the actual values we use, take either the
+                // expr the user specified or, if they didn't
+                // specify something for this field name, create a
+                // path relative to the base (which must have been
+                // supplied, or the IR is internally
+                // inconsistent).
+                let fields: Vec<Lvalue<H>> =
+                    field_names.into_iter()
+                               .map(|n| match fields_map.get(&n) {
+                                   Some(v) => v.clone(),
+                                   None => base.clone().unwrap().field(n)
+                               })
+                               .collect();
+
+                block.and(Rvalue::Aggregate(AggregateKind::Adt(adt_def, variant_index, substs),
+                                            fields))
+            }
+            ExprKind::Literal { .. } => {
+                let constant = this.as_constant(expr);
+                block.and(Rvalue::Constant(constant))
+            }
+            ExprKind::Block { .. } |
+            ExprKind::Match { .. } |
+            ExprKind::If { .. } |
+            ExprKind::Loop { .. } |
+            ExprKind::LogicalOp { .. } |
+            ExprKind::Call { .. } |
+            ExprKind::Field { .. } |
+            ExprKind::Deref { .. } |
+            ExprKind::Index { .. } |
+            ExprKind::VarRef { .. } |
+            ExprKind::SelfRef |
+            ExprKind::Assign { .. } |
+            ExprKind::AssignOp { .. } |
+            ExprKind::Break { .. } |
+            ExprKind::Continue { .. } |
+            ExprKind::Return { .. } |
+            ExprKind::StaticRef { .. } => {
+                // these cannot be evaluated as rvalues, so make an lvalue and then
+                // return that
+                let lvalue = unpack!(block = this.as_lvalue(block, expr));
+                block.and(Rvalue::Use(lvalue))
+            }
+        }
+    }
+
+    /// Compile `expr`, yielding a compile-time constant. Assumes that
+    /// `expr` is a valid compile-time constant!
```

Comment:
@oli-obk I started writing a response and found I had more to say than seemed appropriate for a GitHub comment. See this thread: https://internals.rust-lang.org/t/how-to-handle-constants-in-the-compiler/2543